src/main/java/org/elasticsearch/gateway/GatewayMetaState.java

```diff
@@ -225,9 +218,116 @@ private void pre20Upgrade() throws Exception {
+            }
+            if (writeReason != null) {
+                indicesToWrite.put(shardRouting.index(),
+                        new GatewayMetaState.IndexMetaWriteInfo(indexMetaData, currentIndexMetaData,
+                                writeReason));
+            }
+        }
+        return indicesToWrite.values();
+    }
+
+    /**
+     * Loads the current meta state for each index in the new cluster state and checks if it has to be persisted.
+     * Each index state that is part of the new cluster state will be considered even if this node has no shard of the
+     * index allocated on it. This is only run for master nodes.
+     *
+     * @param event           the cluster state event from which we figure out what is new in each index and should potentially be written
+     * @param currentMetaData the current index state in memory.
+     * @return iterable over all indices states that should be written to disk
+     */
+    public static Iterable<GatewayMetaState.IndexMetaWriteInfo> filterStatesOnMaster(ClusterChangedEvent event, MetaData currentMetaData, MetaStateService metaStateService, ESLogger logger) {
+        Map<String, GatewayMetaState.IndexMetaWriteInfo> indicesToWrite = new HashMap<>();
+        MetaData newMetaData = event.state().metaData();
+        // iterate over all indices but only write if ...
+        for (IndexMetaData indexMetaData : newMetaData) {
+            String writeReason = null;
+            IndexMetaData currentIndexMetaData = maybeLoadIndexState(currentMetaData, indexMetaData, metaStateService, logger);
+            // ... new index or state persistence was disabled?
+            if (currentIndexMetaData == null) {
+                writeReason = "freshly created";
+                // ... version changed
+            } else if (currentIndexMetaData.version() != indexMetaData.version()) {
+                writeReason = "version changed from [" + currentIndexMetaData.version() + "] to [" + indexMetaData.version() + "]";
+            }
+            if (writeReason != null) {
+                indicesToWrite.put(indexMetaData.index(),
+                        new GatewayMetaState.IndexMetaWriteInfo(indexMetaData, currentIndexMetaData,
+                                writeReason));
+            }
+
+        }
+        return indicesToWrite.values();
+    }
+
+    protected static IndexMetaData maybeLoadIndexState(MetaData currentMetaData, IndexMetaData indexMetaData, MetaStateService metaStateService, ESLogger logger) {
+        IndexMetaData currentIndexMetaData = null;
+        if (currentMetaData != null) {
+            currentIndexMetaData = currentMetaData.index(indexMetaData.index());
+        } else {
+            try {
+                currentIndexMetaData = metaStateService.loadIndexState(indexMetaData.index());
+            } catch (IOException e) {
+                logger.debug("failed to load index state ", e);
```

Comment:
can we put a TODO here to maybe bubble this Exception up in the future it seems risky to just swallow it....